ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/xlib/procinfo.cpp

```diff
@@ -0,0 +1,145 @@
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#include "../procinfo.h"
+#include <X11/Xlib.h>
+
+using std::to_string;
+
+static inline int XErrorHandlerImpl(Display *display, XErrorEvent *event) {
+  return 0;
+}
+
+static inline int XIOErrorHandlerImpl(Display *display) {
+  return 0;
+}
+
+static inline unsigned long get_wid_or_pid(Display *display, Window window, bool wid) {
+  XSetErrorHandler(XErrorHandlerImpl);
+  XSetIOErrorHandler(XIOErrorHandlerImpl);
+  unsigned char *prop;
+  unsigned long property;
+  Atom actual_type, filter_atom;
+  int actual_format, status;
+  unsigned long nitems, bytes_after;
+  filter_atom = XInternAtom(display, wid ? "_NET_ACTIVE_WINDOW" : "_NET_WM_PID", True);
+  status = XGetWindowProperty(display, window, filter_atom, 0, 1000, False,
+  AnyPropertyType, &actual_type, &actual_format, &nitems, &bytes_after, &prop);
+  if (status == Success && prop != NULL) {
+    property = prop[0] + (prop[1] << 8) + (prop[2] << 16) + (prop[3] << 24);
+    XFree(prop);
+  }
+  return property;
+}
+
+namespace procinfo {
+
+bool wid_exists(wid_t wid) {
+  XSetErrorHandler(XErrorHandlerImpl);
+  XSetIOErrorHandler(XIOErrorHandlerImpl);
+  bool result = false;
+  Display *display = XOpenDisplay(NULL);
+  Window window = XDefaultRootWindow(display);
+  unsigned char *prop;
+  Atom actual_type, filter_atom;
+  int actual_format, status;
+  unsigned long nitems, bytes_after;
+  filter_atom = XInternAtom(display, "_NET_CLIENT_LIST", True);
+  status = XGetWindowProperty(display, window, filter_atom, 0, 1024, False, 
+  AnyPropertyType, &actual_type, &actual_format, &nitems, &bytes_after, &prop);
+  if (status == Success && prop != NULL && nitems) {
+    if (actual_format == 32) {
+      unsigned long *array = (unsigned long *)prop;
+      for (unsigned i = 0; i < nitems; i++) {
+        if (stoul(wid, nullptr, 10) == (unsigned long)array[i]) {
+          result = true;
+          break;
+        }
+      }
+    }
+    XFree(prop);
+  }
+  XCloseDisplay(display);
+  return result;
+}
+
+window_t window_from_wid(wid_t wid) {
+  return stoull(wid, nullptr, 10);
+}
+
+wid_t wid_from_window(window_t window) {
+  return to_string(reinterpret_cast<unsigned long long>(window));
+}
+
+process_t pid_from_wid(wid_t wid) {
+  XSetErrorHandler(XErrorHandlerImpl);
+  XSetIOErrorHandler(XIOErrorHandlerImpl);

```

Comment:
You seem to repeat these same two lines quite a bit, could possibly extract a helper.